src/ngAnimate/animateQueue.js

```diff
@@ -0,0 +1,558 @@
+    var isAnimatableClassName = !classNameFilter
+              ? function() { return true; }
+              : function(className) {
+                return classNameFilter.test(className);
+              };
+
+    function nextCallbackID() {
+      return '__animateID__' + (callbackCounter++);
+    }
+
+    function findCallbacks(element, event) {
+      // TODO use parentNode.contains
+      var isGeneratedID = false;
+      var id = element.attr('id');
+      if (!element.attr('id')) {
+        isGeneratedID = true;
+        id = nextCallbackID();
+        element.attr('id', id);
+      }
+
+      var matches = [];
+      var entries = callbackRegistry[event];
+      if (entries) {
+        forEach(entries, function(entry) {
+          var node = entry.node;
+          if (node.getAttribute('id') === id || node.querySelector('#' + id)) {
+            matches.push(entry.callback);
+          }
+        });
+      }
+
+      if (isGeneratedID) {
+        element.removeAttr('id');
+      }
+
+      return matches;
+    }
+
+    // $animate.on('enter', element, function() {
+    //
+    // });
+    // $animate.on('enter', '[ng-view]', function() {
+    //
+    // });
+    // element.on('$animate:enter', j
+    function triggerCallback(event, element, phase, data) {
+      $$rAF(function() {
+        forEach(findCallbacks(element, event), function(callback) {
+          callback(element, phase, data);
+        });
+      });
+    }
+
+    return {
+      on: function(event, container, callback) {
+        var node = extractElementNode(container);
+        callbackRegistry[event] = callbackRegistry[event] || [];
+        callbackRegistry[event].push({
+          node: node,
+          callback: callback
+        });
+      },
+
+      off: function(event, container, callback) {
+        var entries = callbackRegistry[event];
+        if (!entries) return;
+
+        switch (arguments.length) {
+          case 1:
+            callbackRegistry[event] = null;
+            break;
+
+          case 2:
+            callbackRegistry[event] = filterFromRegistry(entries, container);
+            break;
+
+          case 3:
+            callbackRegistry[event] = filterFromRegistry(entries, container, callback);
+            break;
+        }
+
+        function filterFromRegistry(list, matchContainer, matchCallback) {
+          var containerNode = matchContainer[0];
+          return list.filter(function(entry) {
+            var isMatch = entry.node === containerNode &&
+                            (!matchCallback || entry.callback === matchCallback);
+            return !isMatch;
+          });
+        }
+      },
+
+      push: function(element, event, options, domOperation) {
+        element = stripCommentsFromElement(element);
+        options = options || {};
+        options.domOperation = domOperation;
+        return queueAnimation(element, event, options);
+      },
+
+      enabled: function(bool) {
```

Comment:
Basically, I think it has 4 signatures:

```
() - global getter
(bool) - global setter
(element) - element getter
(element, bool) - element setter
```

The idea is to make the function body less complex, by showing all possible arguments and if necessary re-assign (preferrably at the top of the function and not is several places inside the body).
IIRC, the same "technique" is used in other places inside Angular (e.g. some `$http` or `$resource` callbacks).

I was thinking something like this:

```js
function (element, bool) {
  var argCount = arguments.length;

  if (argCount === 0) {
    // () - Global getter
    bool = !!animationsEnabled;    
  } else {
    var hasElement = isElement(element);
    
    if (!hasElement) {
      // (bool) - Global setter
      bool = animationsEnabled = !!element;
    } else {
      var node = element.length ? element[0] : element;
      var recordExists = disabledElementsLookup.get(node);

      if (argCount === 1) {
        // (element) - Element getter
        bool = !recordExists;
      } else {
        // (element, bool) - Element setter
        bool = !!bool;
        if (!bool) {
          disabledElementsLookup.put(node, true);
        } else if (recordExists) {
          disabledElementsLookup.remove(node);
        }
      }
    }
  }

  return bool;
}
```

I think the body of the function could be made much less complex this way.